### PR TITLE
osd/PG: skip rollforward when !transaction_applied during append_log()

### PR DESCRIPTION
**Pull request by neha-ojha**
Earlier, we did pg_log.roll_forward(&handler), when !transaction_applied, which advanced the crt and trimmed the entries in rollforward(). Due to this, during _merge_object_divergent_entries(), when we tried to rollback entries, those objects were not found in the backend, and thus we hit this bug http://tracker.ceph.com/issues/36739.
    
With this change, we are advancing the crt value, without deleting the objects, so that _merge_object_divergent_entries() does not fail because of deleted objects.

Fixes: http://tracker.ceph.com/issues/36739
Signed-off-by: Neha Ojha <nojha@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by neha-ojha** [comment]
100 runs of reproducer for http://tracker.ceph.com/issues/36739 - 1 failure tracked in http://tracker.ceph.com/issues/38784
http://pulpito.ceph.com/nojha-2019-03-14_00:36:19-rados:thrash-erasure-code-wip-2-36739-2019-03-13-distro-basic-smithi/

rados: http://pulpito.ceph.com/nojha-2019-03-14_17:40:36-rados-wip-2-36739-2019-03-13-distro-basic-smithi/

**Comment by neha-ojha** [comment]
retest this please

File: src/osd/PG.cc [hunk 1]
```diff
@@ -3827,8 +3827,13 @@ void PG::append_log(
      /* We must be a backfill peer, so it's ok if we apply
       * out-of-turn since we won't be considered when
       * determining a min possible last_update.
+      *
+      * We skip_rollforward() here, which advances the crt, without
+      * doing an actual rollforward. This avoids cleaning up entries
+      * from the backend and we do not end up in a situation, where the
+      * object is deleted before we can _merge_object_divergent_entries().
```
Comment (xiexingguo) [inline src/osd/PG.cc]: So if this turns out to be an async-recovery peer, it is possible we'll leave some obsolete gen objects behind (without being able to delete them again)?

Also the comment above needs a revision...

Comment (neha-ojha) [reply src/osd/PG.cc]: We'll rely on scrub to clean them up, hence the change to _scan_rollback_obs(). During check_local() on the primary, we do identify such obsolete objects, but ignore them https://github.com/ceph/ceph/blob/master/src/osd/PrimaryLogPG.cc#L13553-L13554.

Yes, I'll fix the comment to include async_recovery peer too.

File: src/osd/PG.cc [hunk 2]
```diff
@@ -3824,7 +3824,7 @@ void PG::append_log(
 
   PGLogEntryHandler handler{this, &t};
   if (!transaction_applied) {
-     /* We must be a backfill peer, so it's ok if we apply
+     /* We must be a backfill or async recovery peer, so it's ok if we apply
       * out-of-turn since we won't be considered when
       * determining a min possible last_update.
```
Comment (xiexingguo) [inline src/osd/PG.cc]: We do not exclude async recovery peers when determining a min possible last_update during peering.
I think we should kill it entirely instead.

**Comment by neha-ojha** [comment]
@liewegas ok to merge?

**Comment by Aran85** [comment]
@neha-ojha let's say if the crt always keep consistent with actual rollforward(consider deletion), it is the case 5) of_merge_object_divergent_entries(), because https://github.com/ceph/ceph/blob/affb7d396f76273e885cfdbcd363c1882496726c/src/osd/PGLog.h#L1026 ..

am i missing something?

File: src/osd/PG.cc [hunk 3]
```diff
@@ -4584,12 +4589,12 @@ void PG::_scan_rollback_obs(const vector<ghobject_t> &rollback_obs)
        i != rollback_obs.end();
        ++i) {
     if (i->generation < trimmed_to.version) {
-      osd->clog->error() << "osd." << osd->whoami
-			<< " pg " << info.pgid
-			<< " found obsolete rollback obj "
-			<< *i << " generation < trimmed_to "
-			<< trimmed_to
-			<< "...repaired";
+      dout(10) << __func__ << "osd." << osd->whoami
```
Comment (dzafman) [inline src/osd/PG.cc]: @neha-ojha It would be nice if this were logged as an error for replicated pools.

**Comment by neha-ojha** [comment]
> @neha-ojha let's say if the crt always keep consistent with actual rollforward(consider deletion), it is the case 5) of_merge_object_divergent_entries(), because
> 
> [ceph/src/osd/PGLog.h](https://github.com/ceph/ceph/blob/affb7d396f76273e885cfdbcd363c1882496726c/src/osd/PGLog.h#L1026)
> 
> Line 1026 in [affb7d3](/ceph/ceph/commit/affb7d396f76273e885cfdbcd363c1882496726c)
> 
>  if (!i->can_rollback() || i->version <= olog_can_rollback_to) { 
> ..
> am i missing something?

The bug in this tracker issue indicates that on a backfill target, we roll_forward(), since the transaction has not been applied. This results in advancing crt and also deleting objects, which makes _merge_object_divergent_entries() unhappy later when it tries to access those entries, which are no more present. With this fix, we will avoid hitting such a scenario, by only updating crt and allowing _merge_object_divergent_entries() to rollback whenever possible.

File: src/osd/PG.cc (hunk 3 above)
Comment (neha-ojha) [reply src/osd/PG.cc]: @dzafman agreed, will add it in a follow up PR.

**Comment by Aran85** [comment]
@neha-ojha sorry for asking question before checking tracker ticket and i will make a more detail description:
the 249'1377 is a applied transaction with log, and 249'1378 only update log without transaction but it rollforward the 249'1377,  in next peering we rewind to 249'1376 , gen divergent entries and *mark crt  to 249'1376 *(i think here is the key point), on next _merge_object_divergent_entries(), 249'1378 is after last_backfill, so no need to rollback transaction. but failed on *rollback  249'1377*, because the gen objecte had deleted. i ever have a pr maybe could avoid this, it is https://github.com/ceph/ceph/pull/22503, that will stop rollbacking  249'1377, because change crt to older value will mark the object unrollbackable..

by the way, #27015 is another issue different with here..

**Comment by Aran85** [comment]
@neha-ojha @liewegas I approve that this could resolve the problem here, but do not think it is a better choise.

**Comment by neha-ojha** [comment]
> @neha-ojha sorry for asking question before checking tracker ticket and i will make a more detail description:
> the 249'1377 is a applied transaction with log, and 249'1378 only update log without transaction but it rollforward the 249'1377, in next peering we rewind to 249'1376 , gen divergent entries and *mark crt to 249'1376 *(i think here is the key point), on next _merge_object_divergent_entries(), 249'1378 is after last_backfill, so no need to rollback transaction. but failed on _rollback 249'1377_, because the gen objecte had deleted. i ever have a pr maybe could avoid this, it is #22503, that will stop rollbacking 249'1377, because change crt to older value will mark the object unrollbackable..
> 
> by the way, #27015 is another issue different with here..

@Aran85 I think I understand the scenario you are describing and my original approach to fix this bug  https://github.com/ceph/ceph/pull/26347, tried to address it using https://github.com/ceph/ceph/pull/26347/commits/ee9ff223473029cd22ddbde7823e8d8c546a6f38. The idea was to land in Case 5 instead of 4 in _merge_object_divergent_entries(). However, it also resulted in a scenario, where we update the missing set on the backfill target, but that does not get reflected on the primary, which on activation, tries to send transactions to objects, that aren't present on the backfill target.
Your approach in https://github.com/ceph/ceph/pull/22503 seems to be solving the same problem. I still think we might need these to handle other corner cases, beyond this PR(see https://github.com/ceph/ceph/pull/26347#issuecomment-473391869). 
However, this PR definitely helped get rid of the problem we were seeing in http://tracker.ceph.com/issues/36739, and passes the reproducer test case for this issue.

**Comment by Aran85** [comment]
@neha-ojha i think i understand your meaning, i thinks #26347 is the right approach but not this, i see the problem your descripted, it is not only the problem that #26347 will result, but current master will have the problem, consider the scenario now backfill failure(more like EIO) on ec pool(need some more detail description), so i means we need an approach to *notify peer missing*.. i had impl it on my environment, but in testing..

**Comment by Aran85** [comment]
stopping rollback is better than mark unrollbackable :+1: , so I closed the #22503

**Comment by neha-ojha** [comment]
> @neha-ojha i think i understand your meaning, i thinks #26347 is the right approach but not this, i see the problem your descripted, it is not only the problem that #26347 will result, but current master will have the problem, consider the scenario now backfill failure(more like EIO) on ec pool(need some more detail description), so i means we need an approach to _notify peer missing_.. i had impl it on my environment, but in testing..

Yes, notifying the primary and updating peer_missing seemed to be a much more involved process, but if you have a way out, I'd be happy to take a look at it!

**Comment by Aran85** [comment]
@neha-ojha Excuse me, I rechecked the issue you described, and I'm a little confused..

> where we update the missing set on the backfill target, but that does not get reflected on the primary, 

i think primary could aware the missing on backfill target in get_missing stage  when proc_replica_log(), it is the same as replicated backend, if not, replicated backend will have the same issue?

Would you like to explain a little more? thanks!

**Comment by neha-ojha** [comment]
> @neha-ojha Excuse me, I rechecked the issue you described, and I'm a little confused..
> 
> > where we update the missing set on the backfill target, but that does not get reflected on the primary,
> 
> i think primary could aware the missing on backfill target in get_missing stage when proc_replica_log(), it is the same as replicated backend, if not, replicated backend will have the same issue?
> 
> Would you like to explain a little more? thanks!

@Aran85 See the failures on the test run of this branch https://github.com/ceph/ceph-ci/commits/wip-36739-2019-02-18, http://pulpito.ceph.com/nojha-2019-02-19_18:33:10-rados:thrash-erasure-code-wip-36739-2019-02-18-distro-basic-smithi/. They seem to indicate that the missing populated on the backfill target did not get reflected on the primary, hence it sent complete transactions to missing objects on a backfill target.

**Pull request merged**